zetaclient/evm_client.go

```diff
@@ -1411,15 +1476,38 @@ func (ob *EVMChainClient) GetTxID(nonce uint64) string {
 	return fmt.Sprintf("%d-%s-%d", ob.chain.ChainId, tssAddr, nonce)
 }
 
-func (ob *EVMChainClient) GetBlockByNumberCached(blockNumber uint64) (*ethtypes.Block, error) {
+func (ob *EVMChainClient) GetBlockHeaderCached(blockNumber uint64) (*ethtypes.Header, error) {
+	if header, ok := ob.HeaderCache.Get(blockNumber); ok {
+		return header.(*ethtypes.Header), nil
+	}
+	header, err := ob.evmClient.HeaderByNumber(context.Background(), new(big.Int).SetUint64(blockNumber))
+	if err != nil {
+		return nil, err
+	}
+	ob.HeaderCache.Add(blockNumber, header)
+	return header, nil
+}
+
+// GetBlockByNumberCached get block by number from cache
+// returns block, ethrpc.Block, isFallback, isSkip, error
+func (ob *EVMChainClient) GetBlockByNumberCached(blockNumber uint64) (*ethtypes.Block, *ethrpc.Block, bool, bool, error) {
 	if block, ok := ob.BlockCache.Get(blockNumber); ok {
-		return block.(*ethtypes.Block), nil
+		return block.(*ethtypes.Block), nil, false, false, nil
 	}
 	block, err := ob.evmClient.BlockByNumber(context.Background(), new(big.Int).SetUint64(blockNumber))
 	if err != nil {
-		return nil, err
+		if strings.Contains(err.Error(), "block header indicates no transactions") {
+			return nil, nil, false, true, err // it's ok skip empty block
+		} else if strings.Contains(err.Error(), "transaction type not supported") {
+			rpcBlock, err := ob.rpcClient.EthGetBlockByNumber(blockNumber, true)
+			if err != nil {
+				return nil, nil, true, false, err // fall back on ethRPC but still fail
+			}
+			return nil, rpcBlock, true, false, nil // fall back on ethRPC without error
```

Comment:
Should we also create a cache for the new block type, wdyt ?

Reply:
the [commit](https://github.com/zeta-chain/node/pull/1665/commits/003e851e10e5f76fee146ac76cc4bb564b6c523b)